### ISSUE TITLE
Nest.js has trouble connecting to RabbitMQ with latest amqp-connection-manager

### ISSUE THREAD

**Issue by jwalton** [nestjs/nest#7972]
Hey nest.js folks!

I'm the maintainer of amqp-connection-manager.  I've been making some attempts to modernize the code in this library lately - many years ago it was originally written in CoffeeScript, then migrated to JavaScript.  Recently I merged a number of fixes from contributors, and I migrated the whole codebase to typescript.

The typescript migration happened between 3.3.0 and 3.4.0, and somehow this specific release is [causing some problems](https://github.com/jwalton/node-amqp-connection-manager/issues/172) for some nest.js users.  They're seeing:

```
Error: Channel closed by server: 406 (PRECONDITION-FAILED) with message "PRECONDITION_FAILED - fast reply consumer does not exist"
```

This is a weird error - usually this would be caused by someone trying to post a message with a `replyTo: 'amq.rabbitmq.reply-to'` before setting up a consumer listening on that queue (as per [this](https://www.rabbitmq.com/direct-reply-to.html)).  This sounds like a bug outside of amqp-connection-manager... except all the users claim they can reliably get this to work with 3.3.0 and not with 3.4.0.  The changes I made between 3.3.0 and 3.4.0 are [pretty mild](https://github.com/jwalton/node-amqp-connection-manager/issues/172#issuecomment-906912390), and I don't see how any of them could be somehow causing this error (although, clearly something has changed).  Doubly unfortunately, while lots of users are running into this problem, if I try [this demo repo](https://github.com/rengthp/nestjs-rabbitmq-microservice) it works fine, and no one has been able to generate a minimal test case that reproduces this.

Since I'm not a nest.js user, I'm reaching out to see if anyone familiar with the nest.js RabbitMQ transport has any clever ideas about what could be could be causing this problem or thoughts about ways to recreate it.  (I'm essentially grasping at straws here.  -_-)

Thanks in advance!

**Comment by mohsensaremi** [comment]
Same issue

**Comment by jmcdo29** [comment]
@mohsensaremi are you able to provide a reproduction?

**Comment by mohsensaremi** [comment]
Assume you want to use rabbitMQ for request-response messaging.
Something like this service:
```
@Injectable()
export class IdSequenceService {
  constructor(@Inject('IdSequenceClientProxy') private client: ClientProxy) {}

  generateId() {
    return this.client.send('command.generateId');
  }
}
```

I found out for this to work you have to use 
```
async onApplicationBootstrap() {
    await this.client.connect();
}
```
in where you inject `ClientProxy` (in my example `IdSequenceService`)

without this you will get the following error:
```
UnhandledPromiseRejectionWarning: Error: Channel closed by server: 406 (PRECONDITION-FAILED) with message "PRECONDITION_FAILED - fast reply consumer does not exist"
```

**Comment by mgansler** [comment]
Thanks @mohsensaremi, I can confirm this "fixes" it for me too.

**Comment by jmcdo29** [comment]
There is this mention in the docs:

> The `ClientProxy` is **lazy**. It doesn't initiate a connection immediately. Instead, it will be established before the first microservice call, and then reused across each subsequent call. However, if you want to delay the application bootstrapping process until a connection is established, you can manually initiate a connection using the `ClientProxy` object's `connect()` method inside the `OnApplicationBootstrap` lifecycle hook.

**Comment by mohsensaremi** [comment]
@jmcdo29 
Yes, you are right.
There is this mention in the docs:

> Instead, it will be established before the first microservice call, and then reused across each subsequent call.

The problem is it will throw an Error instead of establishing a connection in the first call.

**Comment by jmcdo29** [comment]
Right, just mentioning why the connect call might be making things work.

If anyone is Anne to provide a repo to work with, I can try to take a look into this and why it _might_ be erroring out. Just a simple git repo that has everything set up

**Comment by kamilmysliwiec** [comment]
@mohsensaremi can you find this file https://github.com/nestjs/nest/blob/master/packages/microservices/client/client-rmq.ts#L82 in your local `node_modules` folder and manually update this condition (the first condition in the `connect()` method) to `if (this.connection)` instead?

**Comment by TrejGun** [comment]
> @mohsensaremi are you able to provide a reproduction?

reproduction is provided in original isuue

**Comment by jmcdo29** [comment]
> > @mohsensaremi are you able to provide a reproduction?
> 
> reproduction is provided in original isuue

> Doubly unfortunately, while lots of users are running into this problem, if I try this demo repo it works fine, and no one has been able to generate a minimal test case that reproduces this.

@TrejGun I assumed that this meant the `demo repo` provided would not trigger the same error, so I was hoping to get a repo that does trigger it

**Comment by TrejGun** [comment]
> > > @mohsensaremi are you able to provide a reproduction?
> > 
> > 
> > reproduction is provided in original isuue
> 
> > Doubly unfortunately, while lots of users are running into this problem, if I try this demo repo it works fine, and no one has been able to generate a minimal test case that reproduces this.
> 
> @TrejGun I assumed that this meant the `demo repo` provided would not trigger the same error, so I was hoping to get a repo that does trigger it

demo repo from the initial post is referring to some code that calls `emit` and the bug is in `send` so there is another repo that calls send. here it is https://github.com/TrejGun/hot-potato-using-nestjs-microservices

**Comment by YuriiKovalenko** [comment]
> @mohsensaremi can you find this file https://github.com/nestjs/nest/blob/master/packages/microservices/client/client-rmq.ts#L82 in your local `node_modules` folder and manually update this condition (the first condition in the `connect()` method) to `if (this.connection)` instead?

it doesn't help

**Comment by MostajeranMohammad** [comment]
when i use rabbit with nest microservice get this 406 PRECONDITION-FAILED error(just in @MesagePattern, @eventPattern is OK)
when use nats with it @messagePattern controller does not receive dataPayload

**Comment by mohsensaremi** [comment]
@YuriiKovalenko @MostajeranMohammad 
I recommend you to use this package instead of the NEstJs built-in RabbitMQ package:
https://github.com/golevelup/nestjs/tree/master/packages/rabbitmq

This package lets you use all the power of RabbitMQ like `exchange` and `routingKey`.

The built-in implementation has minimal capabilities of RabbitMQ.

**Comment by eduscrakozabrus** [comment]
The solution to this problem:
Add in RabbitMQ connection points, services and controllers:

async onApplicationBootstrap() {
    await this.clientTenants.connect();
 
    await this.clientGate.connect();
  }

**Comment by AdrienEtienne** [comment]
Hey!

I have the same issue.
The client seams to send before it consumes the queue `amq.rabbitmq.reply-to`.

I tried to change the line https://github.com/nestjs/nest/blob/5974b6fed12dfe3d0846cb48dcf2cb449bde2557/packages/microservices/client/client-rmq.ts#L196 with

```typescript
setTimeout(() => {
  this.channel.sendToQueue(this.queue, Buffer.from(JSON.stringify(serializedPacket)), {
      replyTo: this.replyQueue,
      correlationId,
      persistent: this.persistent,
  });
}, 500)
```

And it works, so it means the client needs to setup the listeners first when a message is sent.

**Comment by kevin192291** [comment]
Same Issue here, As of Version 7x, this still worked.
In the discord I have posted a link to a minimal clean repository along with the connection to my rabbitMQ server
and un/pw. This is only temporary, so I will not post the link to it here as I will be removing it in the future.

**Comment by mohsensaremi** [comment]
The problem is with `amqp-connection-manager` latest version.
As @jwalton mentioned here:
https://github.com/jwalton/node-amqp-connection-manager/issues/172#issuecomment-906512579

If you install `amqp-connection-manager@3.3.0` it will not throw `406 (PRECONDITION-FAILED)` error and connect to RabbitMQ as normal.

**Comment by MostajeranMohammad** [comment]
i tested this problem with new commit @kamilmysliwiec , 
{
"@nestjs/microservices": "8.0.7",
"amqp-connection-manager": "^3.7.0",
...
}

problem not fixed yet

**Comment by mohsensaremi** [comment]
@MostajeranMohammad 
try this:
```
{
"@nestjs/microservices": "8.0.7",
"amqp-connection-manager": "3.3.0",
...
}
```

**Comment by kamilmysliwiec** [comment]
> i tested this problem with new commit @kamilmysliwiec ,

New commit has not been released yet.

**Comment by kamilmysliwiec** [comment]
Fixed in 8.0.8
